src/Analyzer.PowerShellRuleEngine/PowerShellRuleEngine.cs

```diff
@@ -0,0 +1,115 @@
+            powerShell.AddStatement();
+
+            powerShell.Commands.AddCommand("Test-AzTemplate")
+                .AddParameter("Test", "deploymentTemplate")
+                .AddParameter("TemplatePath", templateFilePath);
+
+            var executionResults = powerShell.Invoke();
+
+            var evaluations = new List<PowerShellRuleEvaluation>();
+
+            foreach (dynamic executionResult in executionResults)
+            {
+                var uniqueErrors = new Dictionary<string, SortedSet<int>>(); // Maps error messages to a sorted set of line numbers
+
+                foreach (dynamic warning in executionResult.Warnings)
+                {
+                    AddErrorToDictionary(warning, ref uniqueErrors);
+                }
+
+                foreach (dynamic error in executionResult.Errors)
+                {
+                    AddErrorToDictionary(error, ref uniqueErrors);
+                }
+
+                foreach (KeyValuePair<string, SortedSet<int>> uniqueError in uniqueErrors)
+                {
+                    var evaluationResults = new List<PowerShellRuleResult>();
+                    foreach (int lineNumber in uniqueError.Value)
+                    {
+                        var evaluationResult = new PowerShellRuleResult(false, lineNumber);
+                        evaluationResults.Add(evaluationResult);
+                    }
+                    var evaluation = new PowerShellRuleEvaluation(executionResult.Name, uniqueError.Key, false, evaluationResults);
+                    evaluations.Add(evaluation);
+                }
+            }
+
+            return evaluations;
+        }
+
+        private static void AddErrorToDictionary(dynamic error, ref Dictionary<string, SortedSet<int>> uniqueErrors)
+        {
+            var lineNumber = 0;
+
+            Type errorType = error.GetType();
+            IEnumerable<PropertyInfo> errorProperties = errorType.GetRuntimeProperties();
+            if (errorProperties.Where(prop => prop.Name == "TargetObject").Any())
+            {
+                if (error.TargetObject is PSObject targetObject && targetObject.Properties["lineNumber"] != null)
+                {
+                    lineNumber = error.TargetObject.lineNumber;
+                }
+            }
+
+            var lineNumberRegex = new Regex(@"\son\sline:\s\d+");
+            var errorMessage = lineNumberRegex.Replace(error.ToString(), string.Empty); 
+
+            if (!uniqueErrors.TryAdd(errorMessage, new SortedSet<int> { lineNumber }))
+            {
+                // errorMessage was already added to the dictionary
+                uniqueErrors[errorMessage].Add(lineNumber);
+            }
+        }
+
+        static void HandleDataAddedInStreams(object newData, ConsoleColor? color = null)
+        {
+            if (color.HasValue) {
+                Console.ForegroundColor = color.Value;
+            }
+
+            Console.WriteLine(newData);
```

Comment:
We don't want to show in the console possible unexpected errors in the PowerShell scripts? I think it would be especially important if people decide to write down their own checks. I guess we could store those strings and return them to the caller, but I don't see a simple way to do that right now, what do you think?

Reply:
We will need to provide a different mechanism when it comes to authoring custom scripts.  Displaying all console output from PS scripts sort of clutters the console, where we want to just output guidance for the user (i.e. something like "this test failed - fix on this line number").  In scenarios where this is consumed by some other library, there may not be a console to output to at all.